tests/test_metrics.py

```diff
@@ -0,0 +1,57 @@
+from unittest import mock
+
+import jax.numpy as jnp
+import pytest
+from jax import random
+
+from blackjax.inference import metrics
+
+KEY = random.PRNGKey(0)
+
+
+@pytest.fixture()
+def patch_jax(monkeypatch):
+    """Patches JNP multiply and dot to determine which one is called"""
+    multiply, dot = mock.MagicMock(), mock.MagicMock()
+    monkeypatch.setattr(metrics.jnp, "multiply", multiply)
+    monkeypatch.setattr(metrics.jnp, "dot", dot)
+    return multiply, dot
+
+
+def test_gaussian_euclidean_ndim_1(patch_jax):
+    """Test Gaussian Euclidean Function returns correct function when ndim is one"""
+    multiply, _ = patch_jax
+    x = jnp.ones(shape=1)
+
+    assert multiply.call_count == 0
+    momentum, velocity = metrics.gaussian_euclidean(x)
+
+    try:
+        momentum(KEY, [1])
+    except RecursionError:
+        pass
+    assert multiply.call_count == 1
+
+
+def test_gaussian_euclidean_ndim_2(patch_jax):
```

Comment:
I learned a lot reading this!

I am wondering if instead of monkey patching the methods we could implement some functional tests? It would be simpler and test the whole chain. The issue with very granular tests is that you end up fighting against them every time you make a small change that doesn't affect the rest of the library or the results.

I tend to only test interfaces; tests are here to guarantee a contract that you make with users and other developers that are interfacing with this code. This goes both ways: the output of your function should be "correct", that includes making sure at initialization that you are not going to make computations with garbage. How about making sure here that:

1. An exception is raised with the wrong number of dimensions;
2. For 1- and 2-dimensional arrays we get values for the momentum and kinetic energy that are correct;

Reply:
First off happy to change it and will do so sometime soon. Anything I say after this is just for conversation.

Dont interface tests suffer from the condition youre mentioning about fighting the testing suite after a small change? If someone makes a chance to a small piece one line of a function now you have to dig through a bunch of integration tests to figure out where it broke, versus with smaller unit tests only the the piece that broke will have an issue?

This is not an argument against integration tests, theyre great for all the reasons you mention. In my experience having both has been helpful so even if two things fail, at least one of them tells me specifically what is failing that helps fix everything else.

Reply:
If your interfaces are well-designed they don't; with a mature library interfaces are less likely to change than what they enclose. I would argue you *want* the tests to break when there is a change of interface rather, but not when you add a numerical optimization somewhere for instance. Unit tests would break on the latter.

I am not against unit tests in general; I find that they are very useful when you are first building the library as it makes you confident that you are moving in the right direction. But eventually they become a burden. Plus when you make a modification later, you know exactly what broke the tests on the interface: the modification you just made.

Reply:
You want to optimize for developer speed and test suite running speed as much as you can. When  you have too many unit tests they tend to get in the way of developers and slow down the CI.